src/pipewire/pipewire.cc

```diff
@@ -159,36 +161,61 @@ void PipeWireOutput::pause(bool pause)
 {
     pw_thread_loop_lock(m_loop);
     pw_stream_set_active(m_stream, !pause);
+    clock_gettime(CLOCK_REALTIME, &m_time);
     pw_thread_loop_unlock(m_loop);
 }
 
 int PipeWireOutput::get_delay()
 {
-    return (m_buffer_at / m_stride + m_frames) * 1000 / m_rate;
+    int buff_time = (((m_buffer.len() + m_buffer_size) / m_stride) * 1000) / m_rate;
+
+    // Get time difference from last buffer fill
+    timespec time_now {};
+    clock_gettime(CLOCK_REALTIME, &time_now);
+
+    int time_diff =
+        (((time_now.tv_sec - m_time.tv_sec) * 1000000000) +
+         (time_now.tv_nsec - m_time.tv_nsec)) / 1000000;
+
+    if (time_diff < 0)
+        time_diff = 0;
+    if (time_diff > buff_time)
+        time_diff = buff_time;
+
+    return buff_time - time_diff;
 }
 
 void PipeWireOutput::drain()
 {
     pw_thread_loop_lock(m_loop);
-    if (m_buffer_at > 0)
-        pw_thread_loop_timed_wait(m_loop, 2);
+
+    int buflen;
+    while ((buflen = m_buffer.len()) > 0)
+    {
+        pw_thread_loop_timed_wait(m_loop, 1);
+        if (buflen <= m_buffer.len())
```

Comment:
This part looks a little suspect. Are we waiting 1ms, and then expecting the buffer level to have decreased? That seems like it will break if PipeWire uses a period greater than 1ms.

Is there no PipeWire API to a period-based (rather than time-based) wait?

Reply:
From PW doc it is waiting 1 second. Checked also PW source code thread-loop.c:
int pw_thread_loop_timed_wait(struct pw_thread_loop *loop, int wait_max_sec) ...
w_thread_loop_get_time(loop,&timeout, wait_max_sec * SPA_NSEC_PER_SEC)

Reply:
Okay that's fine then. Thanks for checking.